qiskit/circuit/singleton.py

```diff
@@ -258,37 +378,118 @@ def __new__(mcs, name, bases, namespace, *, overrides=None, **kwargs):
+        same singleton instance will be returned.  We suggest that you use a tuple of the values of
+        all arguments that can be set while maintaining the singleton nature.
 
-    def to_mutable(self):
-        return self.base_class(_force_mutable=True)
+        Only keys that match the default arguments or arguments given to ``additional_singletons``
+        at class-creation time will actually return singletons; other values will return a standard
+        mutable instance.
+
+        .. note::
+
+            The singleton machinery will handle an unhashable return from this function gracefully
+            by returning a mutable instance.  Subclasses should ensure that their key is hashable in
+            the happy path, but they do not need to manually verify that the user-supplied arguments
+            are hashable.  For example, it's safe to implement this as::
+
+                @staticmethod
+                def _singleton_lookup_key(*args, **kwargs):
+                    return None if kwargs else args
+
+            even though a user might give some unhashable type as one of the ``args``.
+        """
+        return None
+
+
+class _frozenlist(list):
+    __slots__ = ()
+
+    def _reject_mutation(self, *args, **kwargs):
+        raise TypeError("'params' of singletons cannot be mutated")
+
+    append = clear = extend = insert = pop = remove = reverse = sort = _reject_mutation
+    __setitem__ = __delitem__ = __iadd__ = __imul__ = _reject_mutation
+
+
+class _SingletonInstructionOverrides(Instruction):
+    """Overrides for the mutable methods and properties of `Instruction` to make it immutable."""
+
+    __slots__ = ()
+
+    # The split between what's defined here and what's defined in the dynamic `_Singleton` class is
+    # slightly arbitrary, but generally these overrides are for things that are about the nature of
+    # the `Instruction` class itself, while `_Singleton` handles the Python data model and things
+    # that can't be written in terms of the `Instruction` interface (like the overrides of
+    # `base_class` and `to_mutable`).
+
+    @staticmethod
+    def _prepare_singleton_instance(instruction: Instruction):
+        """Class-creation hook point.  Given an instance of the type that these overrides correspond
+        to, this method should ensure that all lazy properties and caches that require mutation to
+        write to are eagerly defined.
+
+        Subclass "overrides" classes can override this method if the user/library-author-facing
+        class they are providing overrides for has more lazy attributes or user-exposed state
+        with interior mutability."""
+        instruction._define()
+        # Some places assume that `params` is a list not a tuple, unfortunately.  The always-raises
+        # `__setattr__` of the classes that inherit this means the `params` setter can't be used to
+        # replace the object.
+        instruction._params = _frozenlist(instruction._params)
```

Comment:
Is there a reason to not just use `tuple(instruction._params)`. Like either way we'll error either way on mutation, is there something doing explicit type checking for `._params` to be a `list`?

Reply:
I'll run some tests with it, but my recollection is that there's hairy bits of `opflow` that do weird stuff like `all_params = x.params + y.params`, which implicitly requires that they're both the same C sequence type.

Reply:
It's fine I guess, from a strict typing perspective we define the `params` attribute as a `list` explicitly so this is correct even if it's a bit less than desirable.

Reply:
For posterity: after running a couple of tests we realised that there's a much bigger reason we'd forgotten about.  Well-behaved code that wants to use `gate.params` for its own purposes (for example creating a new gate) often does `gate.params.copy()` to ensure it's got an owned version, and since `tuple` doesn't have `copy` method (it doesn't make sense on immutable collections), it would have failed everything.